docker-receive/auth.go

```diff
@@ -0,0 +1,47 @@
+package main
+
+import (
+	"net/http"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/docker/distribution/context"
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/docker/distribution/registry/auth"
+	"github.com/flynn/flynn/pkg/httphelper"
+)
+
+func init() {
+	auth.Register("flynn", auth.InitFunc(newAuth))
+}
+
+func newAuth(options map[string]interface{}) (auth.AccessController, error) {
+	return &Auth{key: options["auth_key"].(string)}, nil
+}
+
+type Auth struct {
+	key string
+}
+
+// Authorized implements the auth.AccessController interface and authorizes a
+// request if it is either considered to be internal (e.g. image pulls using
+// docker-receive.discoverd) or includes the correct basic auth key
+func (a *Auth) Authorized(ctx context.Context, accessRecords ...auth.Access) (context.Context, error) {
+	req, err := context.GetRequest(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if !httphelper.IsInternal(req) && !httphelper.IsAuthorized(req, []string{a.key}) {
```

Comment:
Using `IsInternal` is not safe here, this would allow access from the rest of EC2 if security groups are misconfigured (or the Digital Ocean private network, and so on). Should we just put the auth details in the artifact URL for now?